medium-tests/src/test/java/mediumtest/analysis/AnalysisCancellationMediumTests.java

```diff
@@ -0,0 +1,232 @@
+    System.setProperty(ANALYSIS_EXPIRATION_DELAY_PROPERTY_NAME, "PT0S");
+    var filePath = createFile(baseDir, "pom.xml", "");
+    var fileUri = filePath.toUri();
+    var client = harness.newFakeClient()
+      .withInitialFs(CONFIG_SCOPE_ID, baseDir, List.of(new ClientFileDto(fileUri, baseDir.relativize(filePath), CONFIG_SCOPE_ID, false,
+        null, filePath, null, null, true)))
+      .build();
+    var server = harness.newFakeSonarQubeServer()
+      .withPlugin(TestPlugin.XML)
+      .withProject("projectKey")
+      .start();
+    var backend = harness.newBackend()
+      .withSonarQubeConnection(CONNECTION_ID, server)
+      .withBoundConfigScope(CONFIG_SCOPE_ID, CONNECTION_ID, "projectKey")
+      .withExtraEnabledLanguagesInConnectedMode(Language.XML)
+      .start(client);
+
+    var future = backend.getAnalysisService()
+      .analyzeFilesAndTrack(new AnalyzeFilesAndTrackParams(CONFIG_SCOPE_ID, UUID.randomUUID(), List.of(fileUri), Map.of(), false, System.currentTimeMillis()));
+    Thread.sleep(500);
+
+    backend.getConfigurationService().didRemoveConfigurationScope(new DidRemoveConfigurationScopeParams(CONFIG_SCOPE_ID));
+
+    assertThat(future)
+      .failsWithin(2, TimeUnit.SECONDS)
+      .withThrowableThat()
+      .havingCause()
+      .isInstanceOf(ResponseErrorException.class)
+      .asInstanceOf(InstanceOfAssertFactories.type(ResponseErrorException.class))
+      .extracting(ResponseErrorException::getResponseError)
+      .extracting(ResponseError::getCode)
+      .isEqualTo(RequestCancelled.getValue());
+  }
+
+  @SonarLintTest
+  void it_should_cancel_automatic_analysis_when_canceling_task_via_request(SonarLintTestHarness harness, @TempDir Path baseDir)
+    throws InterruptedException, ConfigScopeNotFoundException {
+    var filePath = createFile(baseDir, "pom.xml", "");
+    var fileUri = filePath.toUri();
+    var client = harness.newFakeClient()
+      .withInitialFs(CONFIG_SCOPE_ID, baseDir, List.of(new ClientFileDto(fileUri, baseDir.relativize(filePath), CONFIG_SCOPE_ID, false,
+        null, filePath, null, null, true)))
+      .build();
+    var plugin = newSonarPlugin("xml")
+      .withSensor(WaitingCancellationSensor.class)
+      .generate(baseDir);
+    var backend = harness.newBackend()
+      .withUnboundConfigScope(CONFIG_SCOPE_ID)
+      .withStandaloneEmbeddedPlugin(plugin)
+      .withEnabledLanguageInStandaloneMode(Language.XML)
+      .start(client);
+    var cancelationFilePath = baseDir.resolve("cancellation.result");
+    when(client.getInferredAnalysisProperties(CONFIG_SCOPE_ID, List.of(fileUri))).thenReturn(Map.of(CANCELLATION_FILE_PATH_PROPERTY_NAME, cancelationFilePath.toString()));
+    backend.getFileService().didOpenFile(new DidOpenFileParams(CONFIG_SCOPE_ID, fileUri));
+    Thread.sleep(1000);
+    assertThat(client.getProgressReportsByTaskId().keySet()).hasSize(1);
```

Comment:
We often use these `Thread.sleep()` calls. Can we use await until asserted maybe?

Reply:
Sometimes it's possible, sometimes not. I replaced this specific usage because here it's easy